Pluralize common variant unit names

mkllnk opened the pull request:
#### What? Why?

Closes #4172

<!-- Explain why this change is needed and the solution you propose.
Provide context for others to understand it. -->

This adds the most popular unit names as singular and plural to our
locale for translation. The added Javascript performs a reverse lookup
to find the right singular/plural form of a unit name in that language.

I explained why I chose this solution on the issue: https://github.com/openfoodfoundation/openfoodnetwork/issues/4172#issuecomment-577447644

#### What should we test?
<!-- List which features should be tested and how. -->

Test in English because we haven't translated this yet.

* Go to admin -> products.
* Change a product to items instead of weight or volume.
* Add a unit name like `bunch` or `loaf`.
* Change the quantity including `0`, `1` and `2` of the variant and observe the change in the description.
* Save the changes.
* Go to the shopfront and verify the product description there.

#### Release notes
<!-- Write a line or two to be included in the release notes.
Everything is worth mentioning, because you did it for a reason. -->

Unit descriptions like "1 bunch" or "4 boxes" can now be translated more accurately. The previous logic would sometimes fail to find the right plural, especially in languages other than English.


Changelog Category: Changed

luisramos0 commented on config/locales/en.yml:
```diff
@@ -2721,6 +2721,101 @@ See the %{link} to find out more about %{sitename}'s features and to start using
+  #   puts "      other: \"#{name}s\"";
+  # }
+  unit_names:
+    # Used 379 times.
```
I dont think we need to keep these numbers in this file. The numbers are useful to see what keys to have here but we I think we should remove these "statistics" for here.

mkllnk replied:
Okay, I thought it could be interesting to compare if we update it one day. But I'll remove them.

luisramos0 commented on lib/open_food_network/option_value_namer.rb:
```diff
@@ -72,5 +71,39 @@ def scale_for_unit_value
+
+      private
+
+      def unit_key_lookup
```
I think this is overcomplicated. Why not just load the key/value we need?

mkllnk replied:
Yeah, I need to add some documentation and explain it step by step. So the user enters "fardo". We can't look up "fardo" with I18n because there is no key named that. So we need to look up which key relates to "fardo" and we find `:bale`. Then we can translation `t(:bale, count: 2)` and get "fardos".

Or did you understand this part and you are hinting at something else?

luisramos0 replied:
ok, I didnt understand that. It's not straight forward. Now I understand why this is necessary :+1:

luisramos0 commented on lib/open_food_network/option_value_namer.rb:
```diff
@@ -72,5 +71,39 @@ def scale_for_unit_value
+    # Provides efficient access to unit name inflections.
+    # The singleton property ensures that the init code is run once only.
+    # The OptionValueNamer is instantiated in loops.
+    class I18nUnitNames
```
I think this break the single responsibility rule. We should have this in a separate class.
UnitNamePlural or Pluralizer or I18nUnitNames.
This (including option_value_names) should all go to something like app/services/naming

mkllnk replied:
I agree, I will move it.

mkllnk commented:
@luisramos0 Did you see the js spec covering the pluralizer?

If we move the pluralizer code (js and ruby) to their own classes, should we make the solution independent of variant_unit_name? In that case, we should not use the scope `unit_names` in the locale but something more general like `inflections`.

luisramos0 commented:
yeah, I'd generalize it. but only if it doesnt add even more complexity.

mkllnk commented:
@luisramos0 This is ready for review again.

- I added the missing words used in French.
- I moved the I18n scope from `unit_names` to `inflections` to be more general here.
- I moved the Ruby to its own lib file. I don't think it depends on our application and therefore doesn't need to be a service.
- I realised that I didn't consider the case of people switching locales. In our case that's mainly important for the enterprise user creating variants. They need to create variants in the same language as their locale. Otherwise it won't pluralise. There is no way around this because we can't know which language the user input is and one word can mean different things in different languages. So I added support for multiple locales, almost regretting going down the caching path. It adds so much more complexity.
- And I added more comments to explain everything. What do you think?

luisramos0 commented on lib/open_food_network/i18n_inflections.rb:
```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+module OpenFoodNetwork
+  # Provides efficient access to unit name inflections.
```
this "unit name" should be replaced by "word" as this is no longer related specifically to unit_names. right?

luisramos0 commented on lib/open_food_network/i18n_inflections.rb:
```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+module OpenFoodNetwork
```
I'd love to kill this namespace. First by moving the OFN related services to app/services (most of them). We could create a generic lib folder like lib/utils for these generic classes and existing generic single classes like Locking, Paperclippable, and RackRequestBlocker. What do you think?

mkllnk replied:
Hm, yes, it would be nice to get rid of the namespace. But we need to come up with better namespaces. I don't find `Util` much better. We could use `I18n::Inflections` here but it could look like it's part of the official I18n library. It's just our add-on.

Moving files into their own namespaces would really help to see their independence. And we could create gems if we wanted to share those parts with other projects.

luisramos0 replied:
I see what you mean. Util is not great but I think lib/util is better than lib/OpenFoodNetwork because it makes it obvious the code is not dependent on OFN code.
lib/utils would just be for those small util class (like Inflections).
I think a separate folder under lib would be for a group of related classes that are independent of OFN like we have in Stripe.
In Java the folder lib/common is used for these classes.

Probably a topic for discourse, not this PR :-)

RachL commented:
@mkllnk I staged this on 2 different staging servers, but I still end up with an error when trying to test

> Change a product to items instead of weight or volume.


![image](https://user-images.githubusercontent.com/6525576/74681789-94328480-51c4-11ea-93f4-325774f13b6b.png)


I've tried this on Melon Farm and Suma Whole Foods (Cocoa powder) on Uk staging.

And my latest release test Kimchi Farm 2 on staging FR

luisramos0 commented:
@RachL  when you change to items you must specific the variant name
is that the problem?

luisramos0 commented:
btw PR #4780 will bring meaningful error messages back to this page!

RachL commented:
@luisramos0 ah indeed this is working if I add a unit name before saving. 🤦‍♀ Moving to ready to go!
